lib/src/authentication/client.dart

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// ignore_for_file: import_of_legacy_library_into_null_safe
+
+import 'package:http/http.dart' as http;
+
+import '../http.dart';
+import '../system_cache.dart';
+import 'credential.dart';
+import 'credential_store.dart';
+
+/// This client automatically modifies request to contain required credentials
+/// in request. For example some credentials might add `Authentication` header
+/// to request.
+class _AuthenticationClient extends http.BaseClient {
+  _AuthenticationClient(
+    this._inner, {
+    required this.credential,
+    required this.serverKey,
+  });
+
+  final http.BaseClient _inner;
+  final Credential credential;
+  final String serverKey;
+
+  @override
+  Future<http.StreamedResponse> send(http.BaseRequest request) async {
+    // Let's last time make sure that, we're allowed to use credential for this
+    // request.
+    if (serverKeyMatches(serverKey, request.url.toString())) {
+      await credential.beforeRequest(request);
+    }
+    return _inner.send(request);
+  }
+
+  @override
+  void close() => _inner.close();
+}
+
+Future<T> withAuthenticatedClient<T>(
+  SystemCache systemCache,
+  String server,
+  Future<T> Function(http.Client) fn, {
+  List<String>? alsoMatches,
```

Comment:
This probably needs a better name, I'm not following?

Reply:
Removed `alsoMatches` argument in favour of adding a check for `serverBaseUrl` matching with request.url [here](https://github.com/dart-lang/pub/pull/3007/files#diff-38bbc4e7b857842267a32eefea9dce429375c7045d6509d89437b15bbd70cbffR37).